File: 046-permutations/permutations.cpp
// Given a collection of distinct numbers, return all possible permutations.
//
//
//
// For example,
// [1,2,3] have the following permutations:
//
// [
//   [1,2,3],
//   [1,3,2],
//   [2,1,3],
//   [2,3,1],
//   [3,1,2],
//   [3,2,1]
// ]


class Solution {
public:
    vector<vector<int>> _permute(vector<int>& nums) {
        vector<vector<int>> res;
        res.clear();
        int size=nums.size();
        if(size==0) return res;
        if(size==1){
            vector<int> v={nums[0]};
            res.push_back(v);
            return res;
        }
        for(int i=0;i<size;i++){
            vector<int> new_nums=nums;
            new_nums.erase(new_nums.begin()+i);
            vector<vector<int>> vvi=_permute(new_nums);
            for(auto v:vvi){
                v.push_back(nums[i]);
                res.push_back(v);
            }
        }
        return res;
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> vvi=_permute(nums);
        vector<vector<int>> res;
        res.clear();
        for(auto v:vvi){
            vector<int> vr;
            for(int i=v.size()-1;i>=0;i--)
                vr.push_back(v[i]);
            res.push_back(vr);
        }
        return res;
    }
};

File: 046-permutations/permutations.py
# -*- coding:utf-8 -*-


# Given a collection of distinct integers, return all possible permutations.
#
# Example:
#
#
# Input: [1,2,3]
# Output:
# [
#   [1,2,3],
#   [1,3,2],
#   [2,1,3],
#   [2,3,1],
#   [3,1,2],
#   [3,2,1]
# ]
#
#


class Solution(object):
    def generate(self, nums, got):
        if all(got):
            return []
        n = len(nums)
        res = []
        for i in range(n):
            if got[i]:
                continue
            ng = got[:]
            ng[i] = True
            nxt = self.generate(nums, ng)
            if len(nxt) > 0:
                for p in nxt:
                    res.append([nums[i]] + p)
            else:
                res.append([nums[i]])
        return res
        
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        if n == 0:
            return []
        got = [False for _ in range(n)]
        return self.generate(nums, got)
    
